### PR TITLE
Mixture model

### PR DESCRIPTION
**Pull request by MojiFarmanbar**
- This is an extension of the two-step approach (TwoStepKmeansSampler), however, Gaussian Mixture Models(GMM) method is being used instead of Kmeans. 
- The final samples are selected based on the maximum likelihood
-  see below the results, using the cardinal benchmark on the noisy low-high dimensional sample. 
   -   **noisy-hd-accuracy result**
noisy_hd_accuracy
   - **noisy-hd-noisy-in-selected**
noisy_hd_noisy_in_selected
  - **noisy-ld-accuracy**
noisy_ld_accuracy
  - **noisy-ld-noisy-in-selected** 
noisy_ld_noisy_in_selected

### CODE REVIEW

File: cardinal/clustering.py [hunk 1]
```diff
@@ -79,6 +78,127 @@ def __init__(self, batch_size, **kmeans_args):
...
+        
+        Args:
+            X: Pool of unlabeled samples of shape (n_samples, n_features).
+            sample_weight: Weight of the samples of shape (n_samples),
+                optional.
+
+        Returns:
+            Indices of the selected samples of shape (batch_size).
+        """
+        from scipy.stats import multivariate_normal
```
Comment (AlexandreAbraham) [inline cardinal/clustering.py:119]: Since `scipy` is not an optional package, we can directly import it at the top.

File: cardinal/clustering.py [hunk 2]
```diff
@@ -79,6 +78,127 @@ def __init__(self, batch_size, **kmeans_args):
...
+            try:
+                density = multivariate_normal.logpdf(X, cov=gmm_model.covariances_[i]
+                                                , mean=gmm_model.means_[i]
+                                                , allow_singular=True)
+            except linalg.LinAlgError:
+                cov = gmm_model.covariances_[i]
+                approximated_cov = np.linalg.pinv(cov)
+                density = multivariate_normal.logpdf(X, cov=approximated_cov
+                                                , mean=gmm_model.means_[i]
+                                                , allow_singular=True)
```
Comment (AlexandreAbraham) [inline cardinal/clustering.py:136]: Commas go at the end of the previous line (PEP8)

File: cardinal/clustering.py [hunk 3]
```diff
@@ -79,6 +78,127 @@ def __init__(self, batch_size, **kmeans_args):
...
+        return centers.reshape(gmm_model.n_components,)
+
+class GMMSampler(GMMLikelihoodSampler):
+    """Select samples as highest likelihood to GMM clusters.
+
+    Args:
+        batch_size: Number of samples to draw when predicting.
+    """
+    def __init__(self, batch_size, **gmm_args):
+        check_modules('sklearn', 'clustering.KmeansSampler')
```
Comment (AlexandreAbraham) [inline cardinal/clustering.py:147]: This method just checks the version and raises a meaningful error. Not critical, but it's better to have the right class name.

```suggestion
        check_modules('sklearn', 'clustering.GMMSampler')
```

File: cardinal/clustering.py [hunk 4]
```diff
@@ -79,6 +78,127 @@ def __init__(self, batch_size, **kmeans_args):
...
+        Returns:
+            Indices of the selected samples of shape (batch_size).
+        """
+        from scipy.stats import multivariate_normal
+        if self._not_enough_samples(X):
+            return np.arange(X.shape[0])
+
+        gmm_model = self.clustering_.fit(X)
+        centers= np.empty(shape=(gmm_model.n_components, 1), dtype=int)
+        print(gmm_model.n_components)
```
Comment (AlexandreAbraham) [inline cardinal/clustering.py:125]:
```suggestion
```

File: cardinal/clustering.py [hunk 5]
```diff
@@ -79,6 +78,127 @@ def __init__(self, batch_size, **kmeans_args):
...
+                                                , mean=gmm_model.means_[i]
+                                                , allow_singular=True)
+            except linalg.LinAlgError:
+                cov = gmm_model.covariances_[i]
+                approximated_cov = np.linalg.pinv(cov)
+                density = multivariate_normal.logpdf(X, cov=approximated_cov
+                                                , mean=gmm_model.means_[i]
+                                                , allow_singular=True)
+            centers[i, :] = np.argmax(density) 
+        return centers.reshape(gmm_model.n_components,)
```
Comment (AlexandreAbraham) [inline cardinal/clustering.py:138]: I guess this reshape can be removed by directly creating the array with the right shape but this is minor.

File: cardinal/clustering.py [hunk 6]
```diff
@@ -79,6 +78,127 @@ def __init__(self, batch_size, **kmeans_args):
...
+    Args:
+        batch_size: Number of samples to draw when predicting.
+    """
+    def __init__(self, batch_size, **gmm_args):
+        check_modules('sklearn', 'clustering.KmeansSampler')
+        from sklearn.mixture import GaussianMixture
+
+        if 'n_clusters' in gmm_args:
+            raise ValueError(
+                'You have specified n_clusters={} when creating KMeansSampler.'
```
Comment (AlexandreAbraham) [inline cardinal/clustering.py:152]:
```suggestion
                'You have specified n_clusters={} when creating GMMSampler.'
```

**Comment by AlexandreAbraham** [comment]
Thanks for your contribution the results look really cool! I have left some comments, all cosmetics, so if you do not have time to take care of them let me know and I will merge and do the fix myself. I'll let you know if I have time to run your method in our benchmark to give you the results!

**Review by AlexandreAbraham — CHANGES_REQUESTED** [review]
Cosmetic changes

**Comment by AlexandreAbraham** [comment]
Thanks a lot!
